src/libostree/ostree-gpg-verifier.c

```diff
@@ -304,12 +304,85 @@ _ostree_gpg_verifier_add_key_ascii_file (OstreeGpgVerifier *self,
   g_ptr_array_add (self->key_ascii_files, g_strdup (path));
 }
 
+gboolean
+_ostree_gpg_verifier_add_keyfile_path (OstreeGpgVerifier   *self,
+                                       const char          *path,
+                                       GCancellable        *cancellable,
+                                       GError             **error)
+{
+  g_autoptr(GError) temp_error = NULL;
+  if (!_ostree_gpg_verifier_add_keyfile_dir_at (self, AT_FDCWD, path,
+                                                cancellable, &temp_error))
+    {
+      g_assert (temp_error);
+
+      /* If failed due to not being a directory, add the file as an ascii key. */
+      if (g_error_matches (temp_error, G_IO_ERROR, G_IO_ERROR_NOT_DIRECTORY))
+        {
+          g_clear_error (&temp_error);
+
+          _ostree_gpg_verifier_add_key_ascii_file (self, path);
+        }
+      else
+        {
+          g_propagate_error (error, g_steal_pointer (&temp_error));
+
+          return FALSE;
+        }
+    }
+  return TRUE;
+}
+
+/* Add files that exist one level below the directory at @path as ascii
+ * key files. If @path cannot be opened as a directory,
+ * an error is returned.
+ */
+gboolean
+_ostree_gpg_verifier_add_keyfile_dir_at (OstreeGpgVerifier   *self,
+                                         int                  dfd,
+                                         const char          *path,
+                                         GCancellable        *cancellable,
+                                         GError             **error)
+{
+  g_auto(GLnxDirFdIterator) dfd_iter = { 0, };
+
+  if (!glnx_dirfd_iterator_init_at (dfd, path, FALSE,
+                                    &dfd_iter, error))
+    return FALSE;
+
+    g_autofree char *sep = NULL;
+
+    g_debug ("Adding GPG keyfile dir %s to verifier", path);
+
+    if (!g_str_has_suffix (path, "/"))
+      sep = g_strdup ("/");
+
+    while (TRUE)
+      {
+        struct dirent *dent;
+
+        if (!glnx_dirfd_iterator_next_dent_ensure_dtype (&dfd_iter, &dent,
+                                                         cancellable, error))
+          return FALSE;
+        if (dent == NULL)
+          break;
+
+        if (dent->d_type != DT_REG)
+          continue;
+
+        g_autofree char *iter_path = g_strjoin (sep, path, dent->d_name, NULL);
```

Comment:
You can use `g_build_filename` for this and drop the need for `sep`.

Reply:
Yeah, though strictly speaking...since we're doing fd-relative here, a cleaner approach would be to open the file here, stashing it an another array, and then later using `gpgme_data_new_from_fd()` just like we do for paths.

Or even more strongly, open every file here and then change the verifier to just iterate over the fds.

(This is an optional followup)

Reply:
Ahh, the utility functions are handy. Added this in.

Storing the fds while reading the paths and later iterating over those to add to the verifier sounds clean, saves doing the intermediate work of storing the path. I added a TODO for this.